feat: option for queries to respect accessibility (#1064)
callstack/react-native-testing-library: website/docs/Queries.md

    - [Default state for: `disabled`, `selected`, and `busy` keys](#default-state-for-disabled-selected-and-busy-keys)
    - [Default state for: `checked` and `expanded` keys](#default-state-for-checked-and-expanded-keys)
  - [`ByA11Value`, `ByAccessibilityValue`](#bya11value-byaccessibilityvalue)
- [Common query options](#common-query-options)
  - [`hidden` option](#hidden-option)

>     - [Examples](#examples)

Reviewer: Can we use `**Examples**` instead of `#### Examples` to avoid generating TOC entry for example?

Author: sure

const element = screen.getByA11yValue({ min: 40 });
```

## Common query options

### `hidden` option

All queries have the `hidden` option which enables them to respect accessibility props on components when it is set to `false`. If you set `hidden` to `true`, elements that are normally excluded from the accessibility tree are considered for the query as well.  Currently `hidden` option is set `true` by default, which means that elements hidden from accessibility will be included by default. However, we plan to change the default value to `hidden: false` in the next major release.

You can configure the default value with the [`configure` function](API.md#configure).

An element is considered to be hidden from accessibility based on [`isInaccessible()`](./API.md#isinaccessible) function result.

#### Examples

> Given the following render: 

Reviewer: This looks like a left over from previous version of the doc
```suggestion
```

Author: looks like a git glitch, I have already deleted it

> ```ts

Reviewer: Due to usage of JSX
```suggestion
```tsx
```

const element = screen.getByA11yValue({ min: 40 });
```

> ## Hidden accessibility option

Reviewer: ```suggestion
## `hidden` query option
```

> All queries have the `hidden` option which enables them to respect accessibility props on components when it is set to `false`. If you set `hidden` to `true`, elements that are normally excluded from the accessibility tree are considered for the query as well.  For now if no `hidden` option is defined, `hidden` is set to `true` by default in order to avoid breaking changes. In the near future, we plan to set it to `false` by default so that the default behavior of all queries is to respect accessibility.

Reviewer: ```suggestion
All queries have the `hidden` option which enables them to respect accessibility props on components when it is set to `false`. If you set `hidden` to `true`, elements that are normally excluded from the accessibility tree are considered for the query as well. Currently `hidden` option is set `true` by default, which means that elements hidden from accessibility will be included by default. However, we plan to change the default value to `hidden: false` in the next major release.
```

> An element will be considered inaccessible when it checks one of the following: 
> - it has a style with `display: "none"`
> - it has the prop [`importantForAccessibility='no-hide-descendants'`](https://reactnative.dev/docs/accessibility#importantforaccessibility-android)
> - it has the prop [`accessibilityElementsHidden={true}`](https://reactnative.dev/docs/accessibility#accessibilityelementshidden-ios) 
> - one of its siblings has the prop [`accessibilityViewIsModal={true}`](https://reactnative.dev/docs/accessibility#accessibilityviewismodal-ios)
> - one of its parents is inaccessible due to one of the checks above

Reviewer: ```suggestion
An element are considered to be hidden from accessibility based on [`isInaccessible()`](./API.md#isinaccessible) function result.
```

Author: oh indeed i did not see we had docs for isInaccessible

You can find more documentation about accessibility in React Native [here](https://reactnative.dev/docs/accessibility).

> 

Reviewer: Pls also mention that you can use `configure` to set this option globally

### Examples

#### `display: "none"`

> Given the following render: 
> ```ts
> render(<Text style={{display:"none"}}>I am inaccessible</Text>)
> ```
> 
> Will **find a match**:
> ```ts
> screen.getByText("I am inaccessible", { hidden: true});
> screen.getByText("I am inaccessible");
> ```
> 
> Will **NOT find a match**:
> ```ts
> screen.getByText("I am inaccessible", { hidden: false }); 
> ```

Reviewer: I would suggest compacting the examples as below:

```suggestion
Given the following render: 
```tsx
render(<Text style={{display:"none"}}>I am inaccessible</Text>);

// Ignore hidden elements
expect(screen.queryByText("I am inaccessible", { hidden: false })).toBeFalsy();

// Match hidden elements
expect(screen.getByText("I am inaccessible")).toBeTruthy(); // Defaults to hidden: true for now
expect(screen.getByText("I am inaccessible", { hidden: true})).toBeTruthy();
```

Reviewer: I think single example (display: none) should be fine for now.
